File: leetCode-75/day-4/142/142-linked-list-cycle-2.js
/**
 * Runtime: 89 ms
 * Memory Usage: 45.6 MB
 * Link: https://leetcode.com/problems/linked-list-cycle-ii/submissions/823587529/
 *  //-----------------------------------------------------------------------------
 *  1. Define has set
 *  2. Backup head for the iterations.
 *  3. Interate over the list unit it is not ended
 *  4.     If the hash set contains the node, return the node as result
 *  5      Add the node to hash set
 *  6.     Move current to the next element
 *  7.  Return null if the is no loop.
 *  //-----------------------------------------------------------------------------
 *
 * @param {ListNode} head
 * @return {ListNode}
 */
var detectCycle = function(head) {

    // 1. Define has set
    let set = new Set();

    // 2. Backup head for the iterations.
    var current = head;

    // 3. Interate over the list unit it is not ended
    while(current != null) {

        // 4. If the hash set contains the node, return the node as result
        if(set.has(current)){
            return current;
        }

        // 5. Add the node to hash set
        set.add(current);

        // 6. Move current to the next element
        current = current.next;
    }

    // 7. Return null if the is no loop.
    return null;
};

File: leetCode-75/day-4/142-linked-list-cycle-2/142-linked-list-cycle-2.cs
/**
 * Runtime: 108 ms
 * Memory Usage: 39.6 MB
 * Link: https://leetcode.com/submissions/detail/762972796/
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode DetectCycle(ListNode head) {

        // 1. If the head is null, return null
        if(head == null)
        {
            return null;
        }

        // 2. Create hash set
        var hashSet = new HashSet<ListNode>();

        // 3. Loop while the current is not null
        for(var current = head; current.next != null; current = current.next)
        {
            // 4. Add the current to the hash set, if it fails, return the current;
            if(!hashSet.Add(current))
            {
                return current;
            }
        }

        // 5. Return null as a negative scenario
        return null;
    }
}
